File: 678-valid-parenthesis-string/valid-parenthesis-string.java
class Solution {
    public boolean checkValidString(String s) {
        int leftmin = 0, leftmax = 0;
        for (char c: s.toCharArray()){
            if (c == '('){
                leftmin += 1;
                leftmax += 1;
            }
            else if (c == ')'){
                leftmin -= 1;
                leftmax -= 1;
            }
            else{
                leftmin -= 1;
                leftmax += 1;
            }
            if (leftmax < 0) return false;
            if (leftmin < 0) leftmin = 0;
        }
        return leftmin == 0 ? true : false;
    }
}

File: 678-valid-parenthesis-string/valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        leftmin = leftmax = 0
        for i in s:
            if i == '(':
                leftmin += 1
                leftmax += 1
            elif i == ')':
                leftmin -= 1
                leftmax -= 1
            else:
                leftmin -= 1
                leftmax += 1
            if leftmax < 0:
                return False
            if leftmin < 0:
                leftmin = 0
        return True if leftmin == 0 else False